src/BloomTests/WebLibraryIntegration/BloomS3ClientTests.cs

```diff
@@ -67,6 +72,27 @@ public void UploadBook_Simple_FilesAreOnS3InExpectedDirectory()
 			}
 		}
 
+		[Test]
+		public void UploadBook_ContainsThumbsFile_DontCopyItToS3()
+		{
+			string srcBookPath = MakeBook();
+			AddThumbsFile(srcBookPath);
+			const string excludedFile = "thumbs.db";
+			var storageKeyOfBookFolder = UploadBook(srcBookPath);
+
+			// It's possible that another unit test uploads a book at the same time, so we can't
+			// test for strict equality.
+			Assert.That(_client.GetCountOfAllFilesInBucket(), Is.AtLeast(Directory.GetFiles(srcBookPath).Count()));
+
+			var expectedDir = Path.GetFileName(srcBookPath);
+			foreach (var file in Directory.GetFiles(srcBookPath))
```

Comment:
I would be inclined to omit this loop. There are (presumably) other tests that verify that most of a book's files get uploaded. The point here is NOT to uploaded the forbidden one. And network calls are very heavyweight for unit testing (it's doubtful whether this whole class of tests should be included in unit testing, actually).

In fact, given the (mainly time) cost of uploading a book, I would be inclined to see whether this could be tested as an extra step in some other upload test. That's not good in terms of separating concerns, but in this case it will save appreciable time.

I feel mean complaining about this; it's very good that you've taken the trouble to make a unit test for this new behavior. But we pay the cost of running these tests every time we build.

Another option would be to refactor the upload/download unit tests so that there is a class with a setup that uploads and (and perhaps also downloads) a book with a certain initial state. The individual 'tests' would then just each verify some important fact about the result, and could be very lightweight. The current test, for example, would just verify that the original folder contains a thumbs.db and the end result does not. It might be possible to combine two or more of the existing unit tests into this new class along with this new one (and perhaps break out the several things they verify into a clearer set of distinct tests).

If this seems like too much trouble, maybe make a 'debt' card out of this idea.